adapter/src/hermes/adapter/stdio/stdio.cc

```diff
@@ -0,0 +1,992 @@
+      auto filename = existing.first.st_bkid->GetName();
+      const auto &blob_names = existing.first.st_blobs;
+      if (!blob_names.empty()) {
+        INTERCEPTOR_LIST->hermes_flush_exclusion.insert(filename);
+        LOG(INFO) << "File handler is opened by adapter." << std::endl;
+        hapi::Context ctx;
+        LOG(INFO) << "Adapter flushes " << blob_names.size()
+                  << " blobs to filename:" << filename << "." << std::endl;
+        hermes::api::VBucket file_vbucket(filename, mdm->GetHermes(), true,
+                                          ctx);
+        auto offset_map = std::unordered_map<std::string, hermes::u64>();
+        for (const auto &blob_name : blob_names) {
+          file_vbucket.Link(blob_name, filename, ctx);
+          offset_map.emplace(blob_name, std::stol(blob_name) * kPageSize);
+        }
+        auto trait = hermes::api::FileMappingTrait(filename, offset_map,
+                                                   nullptr, NULL, NULL);
+        file_vbucket.Attach(&trait, ctx);
+        file_vbucket.Delete(ctx);
+        existing.first.st_blobs.clear();
+        INTERCEPTOR_LIST->hermes_flush_exclusion.erase(filename);
+      }
+      ret = 0;
+    }
+  } else {
+    MAP_OR_FAIL(fflush);
+    ret = real_fflush_(fp);
+  }
+  return (ret);
+}
+
+int HERMES_DECL(fclose)(FILE *fp) {
+  int ret;
+  if (hermes::adapter::IsTracked(fp)) {
+    auto mdm = hermes::adapter::Singleton<MetadataManager>::GetInstance();
+    auto existing = mdm->Find(fp);
+    if (existing.second) {
+      LOG(INFO) << "Intercept fclose." << std::endl;
+      LOG(INFO) << "File handler is opened by adapter." << std::endl;
+      if (existing.first.ref_count == 1) {
+        mdm->Delete(fp);
+        hapi::Context ctx;
+        const auto &blob_names = existing.first.st_blobs;
+        if (!blob_names.empty()) {
+          auto filename = existing.first.st_bkid->GetName();
+          LOG(INFO) << "Adapter flushes " << blob_names.size()
+                    << " blobs to filename:" << filename << "." << std::endl;
+          INTERCEPTOR_LIST->hermes_flush_exclusion.insert(filename);
+          hermes::api::VBucket file_vbucket(filename, mdm->GetHermes(), true,
+                                            ctx);
+          auto offset_map = std::unordered_map<std::string, hermes::u64>();
+          for (const auto &blob_name : blob_names) {
+            file_vbucket.Link(blob_name, filename, ctx);
+            offset_map.emplace(blob_name, std::stol(blob_name) * kPageSize);
+          }
+          auto trait = hermes::api::FileMappingTrait(filename, offset_map,
+                                                     nullptr, NULL, NULL);
+          file_vbucket.Attach(&trait, ctx);
+          file_vbucket.Delete(ctx);
+          existing.first.st_blobs.clear();
+          INTERCEPTOR_LIST->hermes_flush_exclusion.erase(filename);
+        }
+        existing.first.st_bkid->Close(ctx);
+        mdm->FinalizeHermes();
```

Comment:
I know why you're doing this, but it seems inefficient to shutdown Hermes on every `fclose`. We should try to think of a better way as we go forward.

Reply:
It won't shut down in MPI case. If it is a pure STDIO code we can initialize and finalize only at open and close respectively.

Reply:
But if a `stdio` program writes 1,000 files for example, we do tons of unnecessary work, right?

Reply:
I agree. However, we don't have any other init and finalize part. For the MPI case I do intercept MPI init and finalize which would be the case for most hpc cases.

Reply:
Plus if users first open multiple times it will only initialize once and finalize at last close. Only if they open and close per file one by one it's bad .